core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -240,7 +240,7 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       val taskDescs = CoarseGrainedSchedulerBackend.this.synchronized {
         // Filter out executors under killing
         val activeExecutors = executorDataMap.filterKeys(executorIsAlive)
-        val workOffers = activeExecutors.map {
+        val workOffers = activeExecutors.filter(_._2.freeCores > 0).map {
```

Comment:
I don't know this code well but the comment above implies that it means to make an offer on all executors?
What is the performance impact anyway?

Reply:
on our cluster there are many executors and tasksets/tasks.  as we know there is a round-robin manner to fill each node with tasks which will be scheduled for each second("spark.scheduler.revive.interval", "1s").   it seams make no sense to schedule tasks to executors which have no free cores.

Reply:
I'm saying there is a comment a few lines above that describes this as a fake offer that explicitly intends to contact all executors. I think we need to figure out if that's still relevant. I don't have git in front of me but check git blame or GitHub to see when this was written?

Reply:
I checked the git log, at the beginning function `makeOffers` do not filter out any resource, maybe it is the reason why the comment is "Make fake resource offers on all executors"
 `// Make fake resource offers on all executors
    def makeOffers() {
      launchTasks(scheduler.resourceOffers(
        executorHost.toArray.map {case (id, host) => new WorkerOffer(id, host, freeCores(id))}))
    }`

Reply:
BTW,  if freeCores < CPUS_PER_TASK, the code as fellow in resourceOffers() is inefficient since `o.cores / CPUS_PER_TASK` = 0
`val tasks = shuffledOffers.map(o => new ArrayBuffer[TaskDescription](o.cores / CPUS_PER_TASK))
 val availableSlots = shuffledOffers.map(o => o.cores / CPUS_PER_TASK).sum`